gapis/api/vulkan/command_builder_utils.go

```diff
@@ -0,0 +1,214 @@
+// Copyright (C) 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vulkan
+
+import (
+	"context"
+
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/memory"
+)
+
+type commandBuilderUtil struct {
+	allocations   *allocationTracker
+	readMemories  []*api.AllocResult
+	writeMemories []*api.AllocResult
+}
+
+func newCommandBuilderUtil(allocations *allocationTracker) *commandBuilderUtil {
+	return &commandBuilderUtil{
+		allocations:   allocations,
+		readMemories:  make([]*api.AllocResult, 0),
+		writeMemories: make([]*api.AllocResult, 0),
+	}
+}
+
+func (util *commandBuilderUtil) allocReadData(ctx context.Context, g *api.GlobalState, v ...interface{}) api.AllocResult {
+	allocateResult := util.allocations.AllocDataOrPanic(ctx, v...)
+	util.readMemories = append(util.readMemories, &allocateResult)
+	rng, id := allocateResult.Data()
+	g.Memory.ApplicationPool().Write(rng.Base, memory.Resource(id, rng.Size))
+	return allocateResult
+}
+
+func (util *commandBuilderUtil) allocWriteData(ctx context.Context, v ...interface{}) api.AllocResult {
+	allocateResult := util.allocations.AllocDataOrPanic(ctx, v...)
+	util.writeMemories = append(util.writeMemories, &allocateResult)
+	return allocateResult
+}
+
+func (util *commandBuilderUtil) observeNewCommand(cmd api.Cmd) {
+	for i := range util.readMemories {
+		cmd.Extras().GetOrAppendObservations().AddRead(util.readMemories[i].Data())
+	}
+	for i := range util.writeMemories {
+		cmd.Extras().GetOrAppendObservations().AddWrite(util.writeMemories[i].Data())
+	}
+}
+
+func createNewCommandPool(
+	ctx context.Context,
+	allocations *allocationTracker,
+	inputState *api.GlobalState,
+	thread uint64,
+	device VkDevice,
+	family uint32) (VkCommandPool, api.Cmd) {
+	commandPoolID := VkCommandPool(newUnusedID(false, func(x uint64) bool {
```

Comment:
Another nit with the existing code, but this could be more obviously written without the lambda as something like

"GetState(inputState).CommandPools().GetNewUnusedID()"